README.md

```diff
@@ -1 +1,124 @@
+
+Stretch goals
+
+7. some key tests
+8. docker packaging
+9. a nice client
+10. magic
+
+## API
+
+Exposed gRPC "Runner" service methods:
+
+```
+Run(ClientID, cmd, args...) (Process, error)
+Stdout(ClientID, Process) (Stream, error)
+Status(ClientID, Process) (Status, error)
+Abort(ClientID, Process) (Status, error)
+```
+
+where: 
+ * the `Run` method that can launch new processes for authenticated callers.
+   * the "role" of the authenticated client will determine the cgroup style the process runs in. This will be extracted from client mTLS cert.
+   * `Run` takes `cmd` and `args` args for the intended linux command and related cmdline parameters and returns a `Process` message
+ * `Process` is an abstracted process ID for the process control calls: `Stdout`, `Status`, and `Abort`. It will be a UUID that maps to the underlying os pid. 
+ * `Status` will return an enum of `COMPLETED|RUNNING|ERROR|ABORTED` and, ideally, the exit code
+
+## AuthN / AuthZ
+
+mTLS client certs will be used for authN. The client cert will map to a set of "roles" (via the `UID` x.509 field) that determines the class of cgroup processes will run under. For authz a simple check that the client `UID` has an entry in the client->assigned_cgroup mapping table will determine if they can launch a process or not. We could extend this such that only some roles can abort a process or attach to it. 
+
+| cgroup template | CPU | Memory | Disk IO | 
+| ------------| --- | ------ | ------ |
+| **good**    | x   | x      | x      |
+| **better**  | x   | x      | x      |
+| **best**    | x   | x      | x      |
+
+| client | assigned cgroup | 
+| -------- | --------------- |
+| **cert-A** | good            |
+| **cert-B** | better        |
+| **cert-C** | best            |
+
+
+## Client UX
+
+The client initially will just be a golang program that runs the server through some paces (just an integration test harness). But in later releases I will try to turn this in to a simple cmdline tool with basic support for each operation. 
+
+```
+> client -cert cert.pem run ls -laR
+aaaa-bbbb-ccc-ddd-eeee
+> client stdout -cert cert.pem -job aaaa-bbbb-ccc-ddd-eeee
+.
+..
+README.md
+> client status -cert cert.pem -job aaaa-bbbb-ccc-ddd-eeee
+COMPLETED
+Exit code: 0
+> client abort -cert cert.pem -job aaaa-bbbb-ccc-ddd-eeee
+already terminated
+```
+
+Beyond the raw client code generated by the gRPC code there will not be a separate "sugar" client library. Although it would be handy for managing the stdout streams. 
+
+## Implementation tradeoffs & security considerations
+
+### Running processes
+
+1. Client makes a mutual mTLS connection to server. 
+1. Server will use `UID` field of client's X.509 cert to determine the client -> role mapping (good, better, best)
+1. Server will create cgroups related to the role of the client
+1. Server will launch process with namespace `SysProcAttr`'s related to the role of the client. `stderr` and `stdout` will be sent to a golang `bytes.Buffer`
```

Comment:
Will stderr and stdout be sent to the same buffer, or separate buffers? (A `bytes.Buffer` is not safe for concurrent writes, so if you're using the same buffer you'll need to add some synchronization)

Also worth noting, you mention both stdout and stderr here, but your RPC is called `Stdout`. Naming is hard indeed :-)

Reply:
I'll likely make a `bytes.Buffer` wrapper with a mutex.  
And I'll fix the naming :)